### PR TITLE
[workflow] Fix workflow recovery issue due to a bug of dynamic output

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Workflow fails to recover to the latest checkpoint due to a bug of dynamic output in storage. For example, 

Step A returns Step B, and Step B returns step C. When step C failed, the workflow recovers from step A instead of step B. This is not a bug of recovery algorithm, but a bug of dynamic output: the latest dynamic output of the workflow does not get
updated, so it points to the checkpoint of step A instead of step B.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
please add descriptions

**Comment by suquark** [comment]
@iycheng This PR is ready for review. These CI errors should not come from this PR. Do you how to start the build w/o updating the PR?

**Comment by iycheng** [comment]
> @iycheng This PR is ready for review. These CI errors should not come from this PR. Do you how to restart the CI build w/o updating the PR?

You should log in to buildkite with github account and then you'll be able to retry this.

File: python/ray/workflow/workflow_storage.py [hunk 1]
```diff
@@ -167,6 +165,10 @@ def save_step_output(self, step_id: StepID, ret: Union[Workflow, Any], *,
                 #     self._put(self._key_step_exception(step_id), exception))
 
         asyncio_run(asyncio.gather(*tasks))
+        if dynamic_output_id:
+            asyncio_run(
+                self._update_dynamic_output(outer_most_step_id,
+                                            dynamic_output_id))
```
Comment (iycheng) [inline python/ray/workflow/workflow_storage.py:171]: Should we put it to tasks and just call `        asyncio_run(asyncio.gather(*tasks))`?

File: python/ray/workflow/workflow_storage.py [hunk 2]
```diff
@@ -167,6 +165,10 @@ def save_step_output(self, step_id: StepID, ret: Union[Workflow, Any], *,
                 #     self._put(self._key_step_exception(step_id), exception))
 
         asyncio_run(asyncio.gather(*tasks))
+        if dynamic_output_id:
```
Comment (iycheng) [inline python/ray/workflow/workflow_storage.py:168]: `if dynamic_output_id is not None`

File: python/ray/workflow/tests/test_basic_workflows.py [hunk 3]
```diff
@@ -282,6 +282,27 @@ def f1(n):
...
+def exponential_fail(k, n):
+    if n > 0:
+        if n < 3:
+            raise Exception("Failed intentionally")
+        return exponential_fail.options(name=f"step_{n}").step(k * 2, n - 1)
+    return k
+
+
+def test_dynamic_output(workflow_start_regular_shared):
+    try:
```
Comment (iycheng) [inline python/ray/workflow/tests/test_basic_workflows.py:295]: add some comments about what's tested here.

File: python/ray/workflow/workflow_storage.py (hunk 1 above)
Comment (suquark) [reply python/ray/workflow/workflow_storage.py:171]: There is a race condition: if we update the dynamic output before finishing checkpointing, then during recovery, the dynamic output would point to a checkpoint that does not exist. Let me comment in the code.

**Review by fishbone — APPROVED** [review]

File: python/ray/workflow/workflow_storage.py [hunk 4]
```diff
@@ -166,8 +164,17 @@ def save_step_output(self, step_id: StepID, ret: Union[Workflow, Any], *,
                 # tasks.append(
                 #     self._put(self._key_step_exception(step_id), exception))
 
+        # Finish checkpointing.
         asyncio_run(asyncio.gather(*tasks))
 
+        # NOTE: if we update the dynamic output before
```
Comment (iycheng) [inline python/ray/workflow/workflow_storage.py:170]: Thanks for the explaination!

**Comment by suquark** [comment]
The CI failures seem not related. I'll merge the PR.

**Pull request merged**